grpc-api/src/main/java/io/deephaven/grpc_api/arrow/FlightServiceGrpcImpl.java

```diff
@@ -0,0 +1,541 @@
+                mi.header = Message.getRootAsMessage(ByteBuffer.wrap(decoder.readRawBytes(size)));
+                continue;
+            } else if (tag == APP_METADATA_TAG) {
+                final int size = decoder.readRawVarint32();
+                mi.app_metadata = BarragePutMetadata.getRootAsBarragePutMetadata(ByteBuffer.wrap(decoder.readRawBytes(size)));
+                continue;
+            } else if (tag == FLIGHT_DESCRIPTOR_TAG) {
+                final int size = decoder.readRawVarint32();
+                final byte[] bytes = decoder.readRawBytes(size);
+                mi.descriptor = Flight.FlightDescriptor.parseFrom(bytes);
+                continue;
+            } else if (tag != BODY_TAG) {
+                log.info().append("Skipping tag: ").append(tag).endl();
+                decoder.skipField(tag);
+                continue;
+            }
+
+            if (mi.inputStream != null) {
+                // latest input stream wins
+                mi.inputStream.close();
+                mi.inputStream = null;
+            }
+
+            final int size = decoder.readRawVarint32();
+
+            //noinspection UnstableApiUsage
+            mi.inputStream = new LittleEndianDataInputStream(new BarrageProtoUtil.ObjectInputStreamAdapter(decoder, size));
+        }
+
+        log.info().append("rcvd do put:")
+                .append(" descriptor=").append(mi.descriptor == null ? "none" : mi.descriptor.toString())
+                .append(" header=").append(mi.header == null ? "none" : MessageHeader.name(mi.header.headerType()))
+                .append(" app_md=").append(mi.app_metadata == null ? "none" : "exists")
+                .append(" body=").append(mi.inputStream == null ? "none" : Integer.toString(mi.inputStream.available()))
+                .endl();
+
+        if (mi.header != null && mi.header.headerType() == MessageHeader.RecordBatch && mi.inputStream == null) {
+            //noinspection UnstableApiUsage
+            mi.inputStream = new LittleEndianDataInputStream(new ByteArrayInputStream(CollectionUtil.ZERO_LENGTH_BYTE_ARRAY));
+        }
+
+        return mi;
+    }
+
+    private static final class MessageInfo {
+        Message header = null;
+        BarragePutMetadata app_metadata = null;
+        Flight.FlightDescriptor descriptor = null;
+        @SuppressWarnings("UnstableApiUsage")
+        LittleEndianDataInputStream inputStream = null;
+    }
+
+    /**
+     * This is a stateful marshaller; a PUT stream begins with its schema.
+     */
+    private static class PutMarshaller extends SingletonLivenessManager implements Closeable {
+        private long nextSeq = 0;
+        private BarrageSourcedTable resultTable;
+        private SessionState.ExportBuilder<Table> resultExport;
+        private final SessionState session;
+        private final StreamObserver<Flight.PutResult> observer;
+        private final TreeMap<Long, MessageInfo> pendingSeq = new TreeMap<>();
```

Comment:
`TreeMap` is overkill, and you can allocate lazily. Maybe use `RAPriQueue` or similar.